File: Assets/My/Scripts/LCallC/LuaCalrefout.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

public class TestRefOut
{
    public static void Func1()
    {

    }
    public static string Func2(string str1, out string str2)
    {
        str1 += "Func2_C1";
        str2 = "Func2_C2";

        return "F2";
    }
    public static string Func3(string str1, ref string str2)
    {
        str1 += "Func3_C1";
        str2 += "Func3_C2";

        return "F3";
    }

    public static string Func4(out string str1, string str2)
    {
        str1 = "Func4_C1";
        str2 += "Func4_C2";

        return "F4";
    }
    public static string Func5(ref string str1, string str2)
    {
        str1 += "Func5_C1";
        str2 += "Func5_C2";

        return "F5";
    }

    public static string Func6(ref string str1, out string str2)
    {
        str1 += "Func6_C1";
        str2 = "Func6_C2";

        return "F6";
    }

    public static string Func7(out string str1, ref string str2)
    {
        str1 = "Func7_C1";
        str2 = "Func7_C2";

        return "F7";
    }
}



public class LuaCalrefout : MonoBehaviour
{
    void Start()
    {
        xLuaEnv.Instance.Dostring("require('LCallC/LuaCalrefout')");
    }

    private void OnDestroy()
    {
        xLuaEnv.Instance.Free();
    }
}

File: Assets/StreamingAssets/Lua/LCallC/LuaCalrefout.lua
print("----Lua RefOut----")


--C# out的返回变量会赋值给out2 ,Lua的返回变量会赋值给out1
local out2
local f1,out1 = CS.TestRefOut.Func2("s1",out2)
print(f1,out1,out2)

local Ref2
local f1,Ref1 = CS.TestRefOut.Func3("s2",Ref2)
print(f1,Ref1,Ref2)

local out2
local f1,out1 = CS.TestRefOut.Func4(out2,"s5")
print(f1,out1,out2)

local Ref2
local f1,Ref1 = CS.TestRefOut.Func5(Ref2,"s7")
print(f1,Ref1,Ref2)

local out,ref
local f1 ,out ,ref = CS.TestRefOut.Func6(out,ref)
print(f1,out,ref)

local out,ref
local f1 ,ref ,out = CS.TestRefOut.Func7(ref,out)
print(f1,out,ref)

--总结 Out ref 会变成多返回值而不是对原来的参数有影响